app_integrations/batcher.py

```diff
@@ -0,0 +1,122 @@
+        """
+        LOGGER.info('Starting batch send of %d logs to the rule processor', len(logs))
+
+        # Try to send all of the logs in one fell swoop
+        if self._send_logs_to_stream_alert(source_function, logs):
+            return
+
+        # Fall back on segmenting the list of logs into multiple requests
+        # if they could not be sent at once
+        return self._segement_and_send(source_function, logs)
+
+    def _segement_and_send(self, source_function, logs):
+        """Protected method for segmenting a list of logs into smaller lists
+        so they conform to the input limit of AWS Lambda
+
+        Args:
+            source_function (str): The app function name from which the logs came
+            logs (list): List of the logs that have been gathered
+        """
+        log_count = len(logs)
+        LOGGER.debug('Segmenting %d logs into subsets', log_count)
+
+        segment_size = int(math.ceil(log_count / 2.0))
+        for index in range(0, log_count, segment_size):
+            subset = logs[index:segment_size + index]
+            # Try to send this current subset to the rule processor
+            # and segment again if they are too large to be sent at once
+            if not self._send_logs_to_stream_alert(source_function, subset):
+                self._segement_and_send(source_function, subset)
+
+        return True
+
+    def _send_logs_to_stream_alert(self, source_function, logs):
+        """Protected method for sending logs to the rule processor lambda
+        function for processing. This performs some size checks before sending.
+
+        Args:
+            source_function (str): The app function name from which the logs came
+            logs (list): List of the logs that have been gathered
+        """
+        # Create a payload to be sent to the rule processor that contains the
+        # service these logs were collected from and the list of logs
+        payload = {'Records': [{'stream_alert_app': source_function, 'gathered_logs': logs}]}
+        payload_json = json.dumps(payload, separators=(',', ':'))
+        if len(payload_json) > MAX_LAMBDA_PAYLOAD_SIZE:
+            LOGGER.debug('Log payload size for %d logs exceeds limit and will be '
+                         'segmented (%d > %d max).', len(logs), len(payload_json),
+                         MAX_LAMBDA_PAYLOAD_SIZE)
+            return False
+
+        LOGGER.debug('Sending %d logs to rule processor with payload size %d',
+                     len(logs), len(payload_json))
+
+        try:
+            response = Batcher.LAMBDA_CLIENT.invoke(
+                FunctionName=self.rp_function,
+                InvocationType='Event',
+                Payload=payload_json,
+                Qualifier='production'
+            )
+
+        except ClientError as err:
+            LOGGER.error('An error occurred while sending logs to '
+                         '\'%s:production\'. Error is: %s',
```

Comment:
why does this have a `:production` suffix?

Reply:
We always invoke the 'production' alias for Lambda functions. See Line 107 above (https://github.com/airbnb/streamalert/pull/323/files/071565d00c53c208aba3ca5e60b6f55bdf19f3ef#diff-dd9370bfe3012b45b5d3cec0e771cc18R107)

Reply:
Also see: https://github.com/airbnb/streamalert/blob/master/stream_alert/rule_processor/sink.py#L80